### PR TITLE
[Feat] RandAugment and AutoAugment

### PR DESCRIPTION
**Pull request by shijianjian**
### Description
Implementation for RandAugment and AutoAugment.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Comment by edgarriba** [comment]
@shijianjian can you fix this here ? https://github.com/kornia/kornia/blob/709cc2e206f384bfacc6f2732203c3d37f09b228/kornia/enhance/adjust.py#L7

**Comment by shijianjian** [comment]
Sure. I will fix it in another PR. Still thinking if to have the RandAugment/AutoAugment included or not.

File: kornia/enhance/adjust.py [hunk 1]
```diff
@@ -508,6 +508,48 @@ def equalize3d(input: torch.Tensor) -> torch.Tensor:
...
+    """
+    _validate_input_dtype(input, accepted_dtypes=[torch.float16, torch.float32, torch.float64], min=0., max=1.)
+    return 1. - input
+
+
+def invert2d(input: torch.Tensor) -> torch.Tensor:
+    r"""Invert color of the input 2D tensor.
+
+    Args:
+        input (torch.Tensor): image tensor with shapes like :math:(C, D, H, W) or :math:(B, C, D, H, W) to equalize.
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:528]: shouldn't be this (B, C, H, W) or (C, H, W) ?

File: kornia/enhance/adjust.py [hunk 2]
```diff
@@ -508,6 +508,48 @@ def equalize3d(input: torch.Tensor) -> torch.Tensor:
...
+
+    Returns:
+        torch.Tensor: Inverted image or images.
+    """
+
+    input = _to_bchw(input)
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:534]: typing

File: kornia/utils/image.py [hunk 3]
```diff
@@ -76,6 +77,58 @@ def _to_bchw(tensor: torch.Tensor, color_channel_num: Optional[int] = None) -> t
...
+
+    if color_channel_num is not None and color_channel_num != 1:
+        channel_list = [0, 1, 2, 3]
+        channel_list.insert(1, channel_list.pop(color_channel_num))
+        tensor = tensor.permute(*channel_list)
+    return tensor
+
+
+def _validate_input_dtype(input: torch.Tensor, accepted_dtypes: List,
+                          min: Optional[float] = None, max: Optional[float] = None) -> None:
```
Comment (edgarriba) [inline kornia/utils/image.py:113]: I suggest to add this functionality into `kornia.testing`, rename and split into to `check_dtype` , `check_min_max` (or similar).

Comment (edgarriba) [reply kornia/utils/image.py:113]: besides, not sure about the cost of computing min max here how much will affect to the total cost of the function

File: test/augmentation/test_augmentation.py [hunk 4]
```diff
@@ -915,6 +915,28 @@ def test_same_on_batch(self, shape, erase_scale_range, aspect_ratio_range):
         print(f._params)
         assert (res[0] == res[1]).all()
 
+    @pytest.mark.parametrize("value", [99., (88., 66., 44.)])
+    def test_value(self, value, device, dtype):
+        torch.manual_seed(10)
+        input = torch.arange(10, device=device, dtype=dtype).reshape(2, 5).repeat(1, 3, 1, 1)
+        f = RandomErasing((0.3, 0.3), (0.3, 0.3), value=value, same_on_batch=False, p=1.)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:922]: BTW, @shijianjian check this comment https://github.com/kornia/kornia/pull/684#discussion_r492263883
we should start trrying this for the functions that we want to overload

File: kornia/augmentation/policy/__init__.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from .randaugment import RandAugment
```
Comment (edgarriba) [inline kornia/augmentation/policy/__init__.py:1]: @shijianjian you mean to move this e.g to `dda` ? /cc @moskomule

File: kornia/enhance/adjust.py [hunk 6]
```diff
@@ -464,14 +463,15 @@ def build_lut(histo, step):
...
+
     Implements Equalize function from PIL using PyTorch ops based on uint8 format:
     https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/autoaugment.py#L352
 
     Args:
         input (torch.Tensor): image tensor with shapes like :math:(C, H, W) or :math:(B, C, H, W) to equalize.
 
     Returns:
-        torch.Tensor: Sharpened image or images.
+        torch.Tensor: Equalized image or images in :math:(B, C, D, H, W).
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:474]: return shape is wrong. I'd just mention that output shape is the same as input

File: kornia/enhance/adjust.py [hunk 7]
```diff
@@ -508,6 +508,48 @@ def equalize3d(input: torch.Tensor) -> torch.Tensor:
...
+    r"""Invert color of the input 2D tensor.
+
+    Args:
+        input (torch.Tensor): image tensor with shapes like :math:(C, D, H, W) or :math:(B, C, D, H, W) to equalize.
+
+    Returns:
+        torch.Tensor: Inverted image or images.
+    """
+    _validate_input_dtype(input, accepted_dtypes=[torch.float16, torch.float32, torch.float64], min=0., max=1.)
+    return 1. - input
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:521]: why not adding an input variable `max_val` so that `return max_val - input` it can be `max_val: float = 1.`
/cc @ducha-aiki

File: kornia/geometry/transform/crop.py [hunk 8]
```diff
@@ -13,6 +13,7 @@
     "bbox_to_mask",
     "infer_box_shape",
     "validate_bboxes",
+    "bbox_generator"
```
Comment (edgarriba) [inline kornia/geometry/transform/crop.py:16]: we should consider having a dedicated module for bboxes. `kornia.geometry.transform.boxes`. I'll open an issue with feature request.

File: kornia/utils/image.py (hunk 3 above)
Comment (edgarriba) [reply kornia/utils/image.py:113]: besides, would be also great to include `kornia.testing.check_shape` to raise errors with undesired shapes list

File: kornia/augmentation/policy/__init__.py (hunk 5 above)
Comment (moskomule) [reply kornia/augmentation/policy/__init__.py:1]: Then, dda is not differentiable data augmentation anymore 😉
Well, but splitting specific augmentation methods to others is good to keep kornia simple, I think.

Comment (shijianjian) [reply kornia/augmentation/policy/__init__.py:1]: @edgarriba I think it is closer to dda functions as it used ```magnitude``` as the parameter. Similar logic to ```dda``` at the current stage.

@moskomule The kornia augmentation APIs are also differentiable (apart from those uint8 based operations). I think we may have another name than ```dda``` like ```meta_augment``` or something? I think it can be:
```python
from kornia import meta_augment
meta_augment.nn (dda.operations)
meta_augment.nn.functional (dda.functional)
meta_augment.policy.RandAugment
meta_augment.policy.FasterAutoAugment.search()
meta_augment.policy.FasterAutoAugment(res.yml)
```

Comment (moskomule) [reply kornia/augmentation/policy/__init__.py:1]: Or `from kornia.augment import search_strategy`, which I think is easier to understand. IIRC, RandAugment also needs search, so we can create more abstract thing like:

```python
from kornia.augment import search_strategy

search_strategy.RandAugment(depth=2).search(export_as='ra.yaml')
search_strategy.FasterAutoAugment(depth=2).search(export_as='faa.yaml')
```

Comment (shijianjian) [reply kornia/augmentation/policy/__init__.py:1]: Indeed, I think ```strategy``` is a better word here. RandAugment/AutoAugment need to search but in a non-differentiable way, also very slow as far as what I know. I am not sure if it is worth implementing. I think at least, I want to include the search function for FasterAutoAugment, and it seems pretty quick (2 or 3 hours) by using your dda repo.


For the naming, I prefer removing the ```search``` part since the search part will be implemented inside each class.
```python
from kornia.augmentation import strategy

faas = strategy.FasterAutoAugmentStrategy(methods=[solarize, etc.], depth=[2, 4, 6])  # search space is specified here
policy = faas.search(temperature=0.1)   # some hparams for searching, like temperature
 # load and continue searching with other hparams
faas2 = FasterAutoAugmentStrategy.load(policy) 
policy2 = faas2.search(temperature=0.5)

strategy.save_strategy(policy2, name='faa.yml')
# load policy for network usage
faa = strategy.FasterAutoAugment(policy='faa.yml')
# or
faa = strategy.FasterAutoAugment(policy=policy2)
```

For the abstraction, I am thinking of
```python
class FasterAutoAugment(AugmentationBase2D):
    def __init__(self, policy):
        if policy.endswith('.yml'):
            self.policy = read_yml(policy)
        else:
            self.policy = policy
    ...
    def forward(self, input):
         ...

class AugmentationStrategy(nn.Module):
    def __init__(self, policy):
        if policy.endswith('.yml'):
            self.policy = read_yml(policy)
        else:
            self.policy = policy
    @classmethod
    def load(cls, policy):
        return cls(policy)
    def search(self, methods, depth, temperature):
        raise NotImplementedError
    def construct(self,):
        # Something like
        # return Augmentation(self.policy)
        raise NotImplementedError
```

@moskomule you are more expert in autoaugment than me. What do you think?

Comment (edgarriba) [reply kornia/augmentation/policy/__init__.py:1]: this is super cool. Does it mean we are creating here an abstract API that can consume any augmentation ? I mean that won't anything specific to faster paper or any other ?

Comment (moskomule) [reply kornia/augmentation/policy/__init__.py:1]: `faas = strategy.FasterAutoAugmentStrategy(methods=[solarize, etc.], depth=[2, 4, 6])  # search space is specified here` is great and allows users to design more flexible augmentation strategies based on their data, resource, etc.
I think it's useful to have `AugmentationStrategy.pretrained(dataset_name)`, especially for AutoAugment, or such expensive ones (cf. https://github.com/tensorflow/tpu/blob/4489103eee24cd612ea3c41958c9320c45080c6f/models/official/efficientnet/autoaugment.py#L36)

Comment (shijianjian) [reply kornia/augmentation/policy/__init__.py:1]: Great. I will modify this PR to align our discussion. Then we may add the FasterAutoAugment on top on that. Will let you both know when I finish it.

File: kornia/enhance/adjust.py (hunk 7 above)
Comment (edgarriba) [reply kornia/enhance/adjust.py:521]: ping @shijianjian to implement this for v0.5
